pkg/querier/worker/util.go

```diff
@@ -122,3 +124,35 @@ func handle(ctx context.Context, request *httpgrpc.HTTPRequest, handler RequestH
 
 	return response
 }
+
+// handleQueryRequest applies unwraps a request and applies it to the handler.
+func handleQueryRequest(ctx context.Context, request *queryrange.QueryRequest, handler RequestHandler) *queryrange.QueryResponse {
+	r, ctx, err := queryrange.QueryRequestUnwrap(ctx, request)
+	if err != nil {
+		return &queryrange.QueryResponse{
+			Status: status.New(codes.Internal, err.Error()).Proto(),
+		}
+	}
+
+	resp, err := handler.Do(ctx, r)
+	if err != nil {
+		if s, ok := status.FromError(err); ok {
+			return &queryrange.QueryResponse{
+				Status: s.Proto(),
+			}
+		}
+
+		return &queryrange.QueryResponse{
+			Status: status.New(codes.Internal, err.Error()).Proto(),
```

Comment:
We need to be careful about the codes. See https://github.com/grafana/loki/pull/11164